### PR TITLE
(MODULES-7120) Refactor Providers / Adapters

### PR DESCRIPTION
**Pull request by Iristyle**
Things remaining for this PR

- [x] Fix / refactor triggers_same? (remove usage of V1::from_manifest_hash)
- [x] Remove V1::from_manifest_hash from win32_taskscheduler_spec.rb
- [x] Remove `Occurrence` class (requires V1::from_manifest_hash to be removed)
- [x] Remove usage of V1::to_manifest_hash, inlining hashes where applicable
- [x] Move code for V1::to_manifest_hash into v1_adapter_spec.rb which should be the only place it's needed
- [x] Remove V2::V1_TYPE_MAP
- [x] Move classes to V2 namespace or Manifest namespace from V1 namespace - updating all callsites

 - Previously the Adapter (v1Adapter / v2Adapter) code returned V1
   trigger representations, which was then consumed by Providers
   (win32_taskscheduler / taskscheduler_api2) that were then
   responsible for converting to manifest hashes.

   Instead, make the adapters reponsible for producing manifest
   hashes so that providers don't have any understanding of the old
   intermediate V1-esque Trigger structure.

   This is the first step towards removing V1 structures.

### CODE REVIEW

File: lib/puppet/provider/scheduled_task/taskscheduler_api2.rb [hunk 1]
```diff
@@ -147,8 +146,7 @@ def trigger=(value)
     end
 
     needed_triggers.each do |trigger_hash|
-      v1trigger = PuppetX::PuppetLabs::ScheduledTask::Trigger::V1.from_manifest_hash(trigger_hash)
-      task.append_trigger(v1trigger)
+      task.append_trigger(trigger_hash)
```
Comment (michaeltlombardi) [inline lib/puppet/provider/scheduled_task/taskscheduler_api2.rb]: after the rename above, should `trigger_hash` be changed to `manifest_hash` or vice versa or left alone?

File: lib/puppet_x/puppetlabs/scheduled_task/trigger.rb [hunk 2]
```diff
@@ -656,6 +710,114 @@ def self.type_from_v1type(v1type)
       V1_TYPE_MAP[v1type]
     end
 
+    def self.to_manifest_hash(iTrigger)
+      # TODO: remove these
+      # v1trigger = V1.from_iTrigger(iTrigger)
+      # return V1.to_manifest_hash(v1trigger)
+
+      # require 'pry'; binding.pry
+      if V1_TYPE_MAP.key(iTrigger.Type).nil?
```
Comment (glennsarti) [inline lib/puppet_x/puppetlabs/scheduled_task/trigger.rb]: As this is in V2...Should this actually error if it doesn't exist in the v1 map?

File: lib/puppet/provider/scheduled_task/taskscheduler_api2.rb (hunk 1 above)
Comment (Iristyle) [reply lib/puppet/provider/scheduled_task/taskscheduler_api2.rb]: Yeah, I think that makes sense.

**Comment by Iristyle** [comment]
Things remaining for this PR

- [ ] Fix / refactor triggers_same? (remove usage of V1::from_manifest_hash)
- [ ] Remove V1::from_manifest_hash from win32_taskscheduler_spec.rb
- [ ] Remove usage of V1::to_manifest_hash, inlining hashes where applicable
- [ ] Move code for V1::to_manifest_hash into v1_adapter_spec.rb which should be the only place it's needed
- [ ] Remove V2::V1_TYPE_MAP

**Comment by Iristyle** [comment]
FYI @michaeltlombardi it looks like the second part of `triggers_same?` in 78dd1f0 breaks idempotency. A manifest like this is enough to demo the problem:

```puppet
scheduled_task { 'doIt':
  command => 'C:\\Windows\\System32\\notepad.exe',
  trigger => [{schedule => 'daily', minutes_interval => '30', start_time => '15:20'}],
}
```

I have a hunch on what's wrong - will keep you posted.

**Comment by Iristyle** [comment]
UPDATE: Figured it out, working on a few more tests to cover those cases @michaeltlombardi

**Comment by Iristyle** [comment]
Also realized that we can't use `>=` @michaeltlombardi because it wasn't added until Ruby 2.3.  The alternative is pretty straightforward, though not as neat looking.

File: lib/puppet_x/puppetlabs/scheduled_task/trigger.rb [hunk 3]
```diff
@@ -65,6 +65,47 @@ def iso8601_datetime_to_local(value)
...
+        }
+      end
+    end
+
+    def self.default_trigger_for(schedule = 'once')
+      now = Time.now
+      type_hash =
+      {
+        'enabled'             => true,
+        # 'end_date'          => '', # or nil?
```
Comment (glennsarti) [inline lib/puppet_x/puppetlabs/scheduled_task/trigger.rb]: Should this be here?

File: spec/unit/puppet_x/puppetlabs/scheduled_task/trigger_spec.rb [hunk 4]
```diff
@@ -175,6 +176,389 @@
...
+        end
+
+        it 'should be able to be specified as a 12-hour "hh:mm pm"' do
+          trigger_hash['start_time'] = '3:13 pm'
+          expect(trigger['start_time']).to eq('15:13')
+        end
+      end
+    end
+
+    describe 'supplyies missing defaults' do
```
Comment (glennsarti) [inline spec/unit/puppet_x/puppetlabs/scheduled_task/trigger_spec.rb]: nit: typo

File: lib/puppet_x/puppetlabs/scheduled_task/trigger.rb (hunk 3 above)
Comment (Iristyle) [reply lib/puppet_x/puppetlabs/scheduled_task/trigger.rb]: Nope, thanks... removed.

File: lib/puppet/provider/scheduled_task/taskscheduler_api2.rb (hunk 1 above)
Comment (Iristyle) [reply lib/puppet/provider/scheduled_task/taskscheduler_api2.rb]: Will save this one for later...

File: lib/puppet_x/puppetlabs/scheduled_task/trigger.rb [hunk 5]
```diff
@@ -346,7 +346,7 @@ def self.bitmask_to_indexes(bitmask)
   end
   end
 
-  class V1
+  class V2
   class Month
     # https://msdn.microsoft.com/en-us/library/windows/desktop/aa381918(v=vs.85).aspx
```
Comment (Iristyle) [inline lib/puppet_x/puppetlabs/scheduled_task/trigger.rb]: This link should be updated to point to the V2 info, not V1

File: lib/puppet_x/puppetlabs/scheduled_task/trigger.rb [hunk 6]
```diff
@@ -307,7 +307,7 @@ def self.bitmask_to_names(bitmask)
   end
   end
 
-  class V1
+  class V2
```
Comment (Iristyle) [inline lib/puppet_x/puppetlabs/scheduled_task/trigger.rb]: Add the MSDN link for V2 info

File: lib/puppet_x/puppetlabs/scheduled_task/trigger.rb [hunk 7]
```diff
@@ -257,7 +257,7 @@ def self.downcase_keys(hash)
     end
   end
 
-  class V1
+  class V2
   class Day
     # https://msdn.microsoft.com/en-us/library/windows/desktop/aa384014(v=vs.85).aspx
```
Comment (Iristyle) [inline lib/puppet_x/puppetlabs/scheduled_task/trigger.rb]: This link should be updated to point to the V2 info, not V1

**Pull request merged**